File: code/memory-c/page.c
#include <stdint.h>

struct Address {
    uint32_t offset_page; // 页号
    uint32_t offset_in_page; // 页内偏移
};

struct PageTable {
    uint32_t length; // 页表长度
    uint32_t element_size; // 一个页的大小，以字节为单位
    uint64_t * data;
};

struct Progress {
    struct PageTable * page_table;
};

uint32_t memory_block_size(struct Address address) {
    uint64_t result = 1;
    for (uint32_t offset_start = 1; offset_start <= address.offset_in_page; offset_start += 1) {
        result *= 2;
    }

    return result;
}

uint64_t progress_find_raw_address(struct Progress * progress, struct Address address) {
    struct PageTable * page_table = progress -> page_table;
    uint32_t block_size = memory_block_size(address);
    uint32_t offset = address.offset_page;

    if (offset >= page_table -> length) {
        // 发生了错误
        exit(1);
    }

    uint64_t offset_memory_lines = (page_table -> data)[offset]; // 内存块号/行号
    uint64_t raw_address = offset_memory_lines * block_size + address.offset_in_page;

    return raw_address;
}

File: code/cache-rs/src/page.rs
/* 
假设，32位机器，以字节为单位，一页 4KB
那有 2^20 页
用 20 位表示页号
用 12 位表示页内偏移  -> 2^12 = 4KB
*/

const PAGE_SIZE: u32 = 4096; //页面大小，4096kb
const PAGE_TABLE_SIZE: usize = 64;
struct Address {
    offset_page: usize, // 页号
    offset_in_page: u32 // 页内偏移
}

impl Address {
    fn from(logic_address: u32) -> Self {
        let offset_page = logic_address / PAGE_SIZE;
        let offset_in_page = logic_address % PAGE_SIZE;

        Self {
            offset_page: offset_page as usize,
            offset_in_page
        }
    }
}


struct PageTable {
    length: u32, // 页表长度
    data: [u32; PAGE_TABLE_SIZE] // 页表数据，对应 map<页号,内存块号>，页号由data数组的索引隐含给出
}

struct Progress {
    page_table: PageTable
}

impl Progress {
    fn find(&self, logic_address: u32) -> Option<u32> {
        let address = Address::from(logic_address);
        if address.offset_page >= self.page_table.length as usize {
            return None;
        }

        let offset_page = address.offset_page;
        let offset_memory_block = self.page_table.data[offset_page]; // 内存块号
        let address_memory_block = offset_memory_block * PAGE_SIZE; // 内存块号地址
        let raw_address = address_memory_block + address.offset_in_page;
        return Some(raw_address);
    }
}
